compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayIndexOfOp.java

```diff
@@ -0,0 +1,236 @@
+import static org.graalvm.compiler.asm.aarch64.AArch64Assembler.ConditionFlag;
+import static org.graalvm.compiler.lir.LIRInstruction.OperandFlag.CONST;
+import static org.graalvm.compiler.lir.LIRInstruction.OperandFlag.REG;
+import static org.graalvm.compiler.lir.LIRInstruction.OperandFlag.STACK;
+
+import org.graalvm.compiler.asm.Label;
+import org.graalvm.compiler.asm.aarch64.AArch64Address;
+import org.graalvm.compiler.asm.aarch64.AArch64Assembler.ShiftType;
+import org.graalvm.compiler.asm.aarch64.AArch64MacroAssembler;
+import org.graalvm.compiler.asm.aarch64.AArch64MacroAssembler.ScratchRegister;
+import org.graalvm.compiler.core.common.LIRKind;
+import org.graalvm.compiler.lir.LIRInstructionClass;
+import org.graalvm.compiler.lir.Opcode;
+import org.graalvm.compiler.lir.asm.CompilationResultBuilder;
+import org.graalvm.compiler.lir.gen.LIRGeneratorTool;
+
+import jdk.vm.ci.code.Register;
+import jdk.vm.ci.meta.JavaKind;
+import jdk.vm.ci.meta.Value;
+
+@Opcode("AArch64_ARRAY_INDEX_OF")
+public final class AArch64ArrayIndexOfOp extends AArch64LIRInstruction {
+    public static final LIRInstructionClass<AArch64ArrayIndexOfOp> TYPE = LIRInstructionClass.create(AArch64ArrayIndexOfOp.class);
+
+    private final boolean isUTF16;
+    private final int arrayBaseOffset;
+
+    @Def({REG}) protected Value resultValue;
+    @Alive({REG}) protected Value arrayPtrValue;
+    @Alive({REG}) protected Value arrayLengthValue;
+    @Alive({REG}) protected Value fromIndexValue;
+    @Alive({REG, STACK, CONST}) protected Value searchValue;
+    @Temp({REG}) protected Value temp1;
+    @Temp({REG}) protected Value temp2;
+    @Temp({REG}) protected Value temp3;
+    @Temp({REG}) protected Value temp4;
+
+    public AArch64ArrayIndexOfOp(int arrayBaseOffset, JavaKind valueKind, LIRGeneratorTool tool, Value result, Value arrayPtr, Value arrayLength, Value fromIndex, Value... searchValues) {
+        super(TYPE);
+        assert byteMode(valueKind) || charMode(valueKind);
+        assert searchValues.length == 1;
+        this.arrayBaseOffset = arrayBaseOffset;
+        this.isUTF16 = charMode(valueKind);
+        resultValue = result;
+        arrayPtrValue = arrayPtr;
+        arrayLengthValue = arrayLength;
+        fromIndexValue = fromIndex;
+        searchValue = searchValues[0];
+        LIRKind archWordKind = LIRKind.value(tool.target().arch.getWordKind());
+        temp1 = tool.newVariable(archWordKind);
+        temp2 = tool.newVariable(archWordKind);
+        temp3 = tool.newVariable(archWordKind);
+        temp4 = tool.newVariable(archWordKind);
+    }
+
+    private static boolean byteMode(JavaKind kind) {
+        return kind == JavaKind.Byte;
+    }
+
+    private static boolean charMode(JavaKind kind) {
+        return kind == JavaKind.Char;
+    }
+
+    @Override
+    public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
```

Comment:
I think this could be improved some, but given I'm about to merge vector instructions, this should be good enough for now. Soon then it can be re-evaluated to potentially use neon instructions.

Too, I haven't looked carefully for correctness yet, but I'll do that after the other refactoring is done